cmd/tuf/app/init.go

```diff
@@ -93,65 +81,100 @@ func InitCmd(ctx context.Context, directory, previous string, targets targetsFla
+	root, err := prepo.GetRootFromStore(store)
 	if err != nil {
 		return err
 	}
-
-	// Add the keys we just provisioned to root and targets
 	keys, err := getKeysFromDir(directory + "/keys")
 	if err != nil {
 		return err
 	}
+	var allRootKeys []*data.PublicKey
 	for _, role := range []string{"root", "targets"} {
+		currentKeyMap := map[string]bool{}
 		for _, tufKey := range keys {
+			currentKeyMap[tufKey.IDs()[0]] = true
 			if err := repo.AddVerificationKeyWithExpiration(role, tufKey, expiration); err != nil {
 				return err
 			}
 		}
-		if err := repo.SetThreshold(role, threshold); err != nil {
-			return err
+		if role == "root" {
+			// This retrieves all the new root keys, but before we revoke any.
+			allRootKeys, err = repo.RootKeys()
+			if err != nil {
+				return err
+			}
 		}
-	}
-
-	// Revoke old root keys used for snapshot and timestamp and roles.
-	for _, role := range []string{"snapshot", "timestamp"} {
-		for _, tufKey := range keys {
-			// Revoke the offline root keys used in root signing v1 for snapshot and timestamp role.
-			// Revocation is done by specifying one of it's key IDs.
-			// The snapshot and timestamp roles in v2+ will be based on online signers on GCP KMS to
-			// facilitate staging project delegations between root signing events..
-			if err := repo.RevokeKey(role, tufKey.IDs()[0]); err != nil {
-				return errors.Wrap(err, "error revoking key")
+		// Revoke any keys that we've rotated out
+		oldKeys, ok := root.Roles[role]
+		if ok {
+			for _, oldKeyID := range oldKeys.KeyIDs {
+				if _, ok := currentKeyMap[oldKeyID]; !ok {
+					if err := repo.RevokeKey(role, oldKeyID); err != nil {
+						return err
+					}
+				}
 			}
 		}
-		if err := repo.SetThreshold(role, 1); err != nil {
+		if err := repo.SetThreshold(role, threshold); err != nil {
 			return err
 		}
 	}
-	// Add online snapshot and timestamp keys with a shorter expiration of three weeks.
+
+	// Revoke old root keys used for snapshot and timestamp and roles and add new keys.
 	for role, keyRef := range map[string]string{"snapshot": snapshotRef, "timestamp": timestampRef} {
-		// Add new online keys
 		signerKey, err := pkeys.GetKmsSigningKey(ctx, keyRef)
 		if err != nil {
 			return err
 		}
-		// Sets a three week (21 days) expiration.
-		if err := repo.AddVerificationKeyWithExpiration(role, signerKey.Key, time.Now().AddDate(0, 0, 21).UTC()); err != nil {
+
+		currentKeys, ok := root.Roles[role]
+		if ok {
+			for _, oldKeyID := range currentKeys.KeyIDs {
+				if oldKeyID == signerKey.Key.IDs()[0] {
+					// This is the signers we want to keep.
+					continue
+				}
+				// Revoke the offline root keys used in root signing v1 for snapshot and timestamp role.
+				// Revocation is done by specifying one of it's key IDs.
+				// The snapshot and timestamp roles in v2+ will be based on online signers on GCP KMS to
+				// facilitate staging project delegations between root signing events..
```

Comment:
If this is meant to be a one-time step, should this live outside of the init step?

Reply:
Hmmmm I guess so -- we could also just remove this entirely.